Detour/Source/DetourNavMeshQuery.cpp

```diff
@@ -998,106 +1106,207 @@ dtStatus dtNavMeshQuery::queryPolygons(const float* center, const float* halfExt
 								  const dtQueryFilter* filter,
-								  dtPolyRef* path, int* pathCount, const int maxPath) const
+								  dtPolyRef* path, int* pathCount, const int maxPath,
+								  dtQueryPathState* state /* = 0 */) const
+{
+	return findMultiStartPath(1, &startRef, endRef, startPos, endPos, filter,
+			path, pathCount, maxPath, state);
+}
+
+/// @par
+///
+/// If the end polygon cannot be reached through the navigation graph,
+/// the last polygon in the path will be the nearest the end polygon.
+///
+/// If the path array is to small to hold the full result, it will be filled as
+/// far as possible from the start polygon toward the end polygon.
+///
+/// The start and end positions are used to calculate traversal costs.
+/// (The y-values impact the result.)
+///
+dtStatus dtNavMeshQuery::findMultiStartPath(const int numStarts,
+                                  const dtPolyRef* startRefs,
+		                          dtPolyRef endRef,
+								  const float* startPoses,
+								  const float* endPos,
+								  const dtQueryFilter* filter,
+								  dtPolyRef* path, int* pathCount,
+								  const int maxPath,
+								  dtQueryPathState* state /* = 0 */) const
 {
 	dtAssert(m_nav);
-	dtAssert(m_nodePool);
-	dtAssert(m_openList);
-	
+
 	if (pathCount)
 		*pathCount = 0;
-	
+
 	// Validate input
-	if (!m_nav->isValidPolyRef(startRef) || !m_nav->isValidPolyRef(endRef) ||
-		!startPos || !endPos || !filter || maxPath <= 0 || !path || !pathCount)
+	if (numStarts <= 0
+		|| !startRefs
+		|| !m_nav->isValidPolyRef(endRef)
+		|| startPoses == 0
+		|| !endPos || !filter || maxPath <= 0 || !path || !pathCount)
 		return DT_FAILURE | DT_INVALID_PARAM;
 
-	if (startRef == endRef)
+	bool allStartRefsValid = true;
+	for (int i = 0; i < numStarts; ++i)
+		if (!m_nav->isValidPolyRef(startRefs[i]))
+			allStartRefsValid = false;
+
+	if (!allStartRefsValid)
+		return DT_FAILURE | DT_INVALID_PARAM;
+
+
 	{
-		path[0] = startRef;
-		*pathCount = 1;
-		return DT_SUCCESS;
+		bool startHasEnd = false;
+		for (int i = 0; i < numStarts; ++i)
+			if (startRefs[i] == endRef)
+				startHasEnd = true;
+
+		if (startHasEnd)
+		{
+			path[0] = endRef;
+			*pathCount = 1;
+			return DT_SUCCESS;
+		}
 	}
-	
-	m_nodePool->clear();
-	m_openList->clear();
-	
-	dtNode* startNode = m_nodePool->getNode(startRef);
-	dtVcopy(startNode->pos, startPos);
-	startNode->pidx = 0;
-	startNode->cost = 0;
-	startNode->total = dtVdist(startPos, endPos) * H_SCALE;
-	startNode->id = startRef;
-	startNode->flags = DT_NODE_OPEN;
-	m_openList->push(startNode);
-	
-	dtNode* lastBestNode = startNode;
-	float lastBestNodeCost = startNode->total;
-	
+
+	dtQueryPathState* ss;
+	if (!state)
+	{
+		dtAssert(m_nodePool);
+		dtAssert(m_openList);
+
+		ss = new (dtAlloc(sizeof(dtQueryPathState), DT_ALLOC_TEMP)) dtQueryPathState;
+		ss->m_nodePool = m_nodePool;
+		ss->m_openList = m_openList;
+	}
+	else
+	{
+		dtAssert(state->m_nodePool);
+		dtAssert(state->m_openList);
+		ss = state;
```

Comment:
If `state` get destructed `ss` became invalid. That's good to have ownership in such cases or deep copy.